releasenotes/notes/50328.yaml

```diff
@@ -0,0 +1,18 @@
+apiVersion: release-notes/v2
+kind: bug-fix
+area: installation
+issues:
+  - 30393
+  - 42792
+releaseNotes:
+- |
+  **Fixed** Allow for re-executions of istio-iptables by skipping apply step if existing rules are compatible.
+---
+apiVersion: release-notes/v2
+kind: feature
+area: networking
+releaseNotes:
+- |
+  **Added** `--cleanup-only` option to istio-iptables to perform a cleanup of chains/rules without creating new ones.
```

Comment:
Why do we have `istio-clean-iptables` and `istio-iptables --cleanup-only`? What is the difference?

Reply:
Both utilities clean iptables, but their approaches differ.
The `istio-clean-iptables` utility (used only in the VM service as far as I know), come with its own CLI that has only a subset of the arguments of the `istio-iptables` args.
On top of that, the `istio-clean-iptables` cmd relies on a set of hardcoded rules and chains to perform the cleanup. This approach requires attention from developers that everything that is added in istio-iptables is also properly cleaned by `istio-clean-iptables`. This creates the risk that clean-iptables and istio-iptables will diverge over time, which is enabled by the fact that the effectivess of clean-iptables is not tested.

For example I see that https://github.com/istio/istio/pull/41431 added the cleaning of non-jump rules in non-Istio chains which were previously missing in istio-clean-iptables.

In contrast, istio-iptables with the `--cleanup-only` option performs the cleaning by reversing/undoing the rules that istio-iptables would have applied if the cleanup flag was not used.
The rules are not hardcoded and they are generated at runtime by the reverse function, also `--cleanup-only` is tested to make sure that the resulting state is clean.
This "dynamic" approach allows to easily support all possible configurations.

I feel that it would be great to deprecate the `istio-clean-iptables` in the future and update the VM to just use the new cleanup flag.

Reply:
> I feel that it would be great to deprecate the istio-clean-iptables in the future and update the VM to just use the new cleanup flag.

+ 1 on that, we got a bug against `istio-clean-iptables` recently anyway that this would have fixed: https://github.com/istio/istio/issues/52835